Towards tr4

jprotze opened the pull request:
Current state of OMPT implementation update towards TR4 of OpenMP spec:

http://www.openmp.org/wp-content/uploads/openmp-tr4.pdf

jmellorcrummey commented:
We have been managing master by keeping it in sync with LLVM's OpenMP from svn. I think this patch should be submitted as a patch to LLVM OpenMP through Phabricator instead of a pull request here.

jprotze commented:
Hi John,

the intention of this pull-request is to have a easily available diff to 
review the changes.
I'll also write a mail to the list directing to this pull-request.

We will create a patch at Phabricator once we are satisfied with the 
changes.

- Joachim

Am 18.01.2017 um 16:09 schrieb John Mellor-Crummey:
> We have been managing master by keeping it in sync with LLVM's OpenMP
> from svn. I think this patch should be submitted as a patch to LLVM
> OpenMP through Phabricator instead of a pull request here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMPToolsInterface/LLVM-openmp/pull/2#issuecomment-273501002>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGSnhhn6OrMUwpWArsy790GIJSWStkoCks5rTis5gaJpZM4Lm8zo>.
>


-- 
Dipl.-Inf. Joachim Protze

IT Center
Group: High Performance Computing
Division: Computational Science and Engineering
RWTH Aachen University
Seffenter Weg 23
D 52074  Aachen (Germany)
Tel: +49 241 80- 24765
Fax: +49 241 80-624765
protze@itc.rwth-aachen.de
www.itc.rwth-aachen.de

sconvent committed:
fix return addresses for loop_begin callbacks using GOMP interface

sconvent committed:
fix frame addresses in task create callback when using GOMP interface

sconvent committed:
completed return address implementation for ompt_callback_parallel_begin

sconvent committed:
completed return address implementation for ompt_callback_parallel_end

sconvent committed:
completed return address implementation for ompt_callback_master

sconvent committed:
completed return address implementation for ompt_callback_flush

jprotze committed:
Fix segfault because pId is set to NULL above. Fix issue #14

sconvent committed:
fix features for testcases

sconvent committed:
fix parallel/nested_lwt testcase

sconvent committed:
restructure macros for OMPT return addresses

sconvent committed:
Merge branch 'towards_tr4' of https://git.rwth-aachen.de/OpenMPTools/LLVM-openmp into towards_tr4

sconvent committed:
replace workshare function pointer with return address in callbacks and tests

sconvent committed:
set parallel_data.ptr to NULL before parallel_begin callback
add nullcheck for parallel_data, task_data, new_task_data and thread_data to tests

jprotze committed:
Update README

omalyshe committed:
Fix build without ompt - some OMPT_SUPPORT guards were missed

jprotze committed:
Merge pull request #15 from omalyshe/towards_tr4

Fix build without ompt

sconvent committed:
enable use of OMP_TOOL_LIBRARIES for single library

sconvent committed:
Merge branch 'towards_tr4' of https://git.rwth-aachen.de/OpenMPTools/LLVM-openmp into towards_tr4

omalyshe committed:
Support OMPT callbacks for multiple roots

jprotze committed:
Merge pull request #17 from omalyshe/towards_tr4

Support OMPT callbacks for multiple roots

jprotze committed:
Merge branch 'towards_tr4' of https://github.com/OpenMPToolsInterface/LLVM-openmp into towards_tr4

jprotze committed:
Adding testcase for OMPT pthread/C++ thread interoperability (pull request #17)

sconvent committed:
remove ompt support for runtime version 45

hansangbae committed:
Implement issue #4, activating a tool

This change addresses issue #4 for UNIX-like platforms.

hansangbae committed:
Implement issue #4, activating a tool

This change addresses issue #4 for UNIX-like platforms.

sconvent committed:
move ompt_control_tool from version 45 to version 50

sconvent committed:
only write to ompt return address variables if they are NULL

sconvent committed:
test omp_control_tool with values other than 0

sconvent committed:
store reenter_frame address in omp_control_tool

jprotze committed:
Merge pull request #18 from hansangbae/towards_tr4

Implement issue #4 Activating a Tool

jprotze committed:
Merge branch 'towards_tr4' of git.rwth-aachen.de:OpenMPTools/LLVM-openmp into towards_tr4

hansangbae committed:
Update thread-state tracking (mostly renaming) - Issue #7

sconvent committed:
make sure omp_control_tool is present even without OMPT_SUPPORT

sconvent committed:
use colon-separated entries in OMP_TOOL_LIBRARIES env var

sconvent committed:
add testcase for OMP_TOOL_LIBRARIES

sconvent committed:
add testcase for omp_control_tool reenter_frame_address

sconvent committed:
adapt ompt_data_none and ompt_data_none to spec and remove old constants

hansangbae committed:
Define mutex implementations - Issue #8

hansangbae committed:
Merge branch 'towards_tr4' of https://github.com/OpenMPToolsInterface/LLVM-openmp into towards_tr4

sconvent committed:
adapt ompt_get_parallel_info and ompt_get_task_info to newest spec

sconvent committed:
remove ompt_callback_lock_init_t and ompt_callback_lock_destroy_t

sconvent committed:
update ompt_task_type_t

hansangbae committed:
Fix incorrect mutex kind in __kmpc_init_nest_lock

sconvent committed:
fix double declaration of ompt_task_type_t

jprotze committed:
Include stddef.h in ompt.h for size_t

sconvent committed:
remove parallel_data from ompt_thread_info_t

sconvent committed:
use new ompt_task_type_t and adapt tests

jprotze committed:
Merge PR#19: Updated thread state tracking, added definitions for mutex implementation.

This change set partially addresses Issue #7 and #8.
For #7, it is mostly name changes, and I didn't touch any other things.
For #8, I wasn't sure what would be an acceptable categorization method, so just think of it as a proposal. Only three implementation types are defined - spin, queuing, speculative.

* speculative - employs HW/SW support for speculative execution
* queuing - enforces fairness
* spin - all other spin-based ones

Fixed Conflicts:
	runtime/src/include/45/ompt.h.var
	runtime/src/kmp_barrier.cpp
	runtime/src/kmp_runtime.cpp

jprotze committed:
We removed ompt-header files from builds of versions 3.0, 4.0 and 4.5, so fail for OMPT + Version < 5.0

sconvent committed:
fix minor error in callback.h

sconvent committed:
add task_type testcase

sconvent committed:
use different separation symbol for task_type to be more consistent

sconvent committed:
parallel_data argument is NULL on implicit-task-end

sconvent committed:
renaming, cleanup and replacing ompt_thread_data_t,ompt_parallel_data_t,ompt_task_data_t with ompt_data_t

sconvent committed:
fix linking error when OMPT_SUPPORT is disabled

hansangbae committed:
Define and use mutex implementation types

sconvent committed:
set parallel_data to NULL on barrier end after parallel_region

sconvent committed:
fix retrieval of task_type info

sconvent committed:
adapt testcases to fixed task_type output

jprotze committed:
For omp_control_tool, the function call is the runtime entry point. So update frame pointer there.

sconvent committed:
fix task_type for final and undeferred tasks

sconvent committed:
store and load correct return address on task create with dependences

sconvent committed:
add task_type_serialized testcase

sconvent committed:
cleanup

sconvent committed:
also print OMP_TOOL_LIBRARIES when OMP_DISPLAY_ENV is set

jprotze committed:
Fix uninitialized value issue for untied tasks

sconvent committed:
enable tests using untied tasks

sconvent committed:
fix control_tool test

sconvent committed:
fix task_type testcases

sconvent committed:
output for undefined OMP_TOOL_LIBRARIES

sconvent committed:
add barrier testcases

sconvent committed:
add sync_region_wait callbackk for taskwait and add corresponding testcase

sconvent committed:
add taskgroup testcase

sconvent committed:
add test_lock testcase and fix callback for test_nest_lock

sconvent committed:
adapt testcases so that gcc can compile them as well

sconvent committed:
fix untied task flag for GOMP interface

jprotze committed:
Merge branch 'towards_tr4' into towards_tr4

jprotze committed:
Merge pull request #20 from hansangbae/towards_tr4

Define and use mutex implementation types
Solves issue #8

jprotze committed:
Remove duplicate definition

jprotze committed:
Make config guards consistent for implementation and usage in settings.

sconvent committed:
fix parallel_data=NULL in implicit_task_end callback for serialized parallel regions

sconvent committed:
initialize parallal_data for implicit parallel region on initial task creation; update testcases accordingly

sconvent committed:
fix barrier/single testcase to avoid unwanted optimizations with gcc

sconvent committed:
fix test_lock testcase

sconvent committed:
add sync_region_wait callbacks for taskgroup and add testcase

sconvent committed:
delete barrier/sections_nowait testcase

sconvent committed:
split up mutex testcase

jprotze committed:
Guard ompt include

hansangbae committed:
Declare omp_control_tool in Fortran header

sconvent committed:
split up test_lock testcase

jprotze committed:
Remove unused old OMPT code - was commented out before

jprotze committed:
Tool data ids are initilized to 0. Remove old functions with option to init with unique ids.

sconvent committed:
restructure FileCheck prefixes to be able to toggle testing of sync_region_wait callbacks

jprotze committed:
Fix some return addresses for implicit barriers of workers

sconvent committed:
reorder substitutions in lit.cfg

jprotze committed:
Store codeptr of parallel teams for use in barrier and parallel end

sconvent committed:
test return addresses in several callbacks

jprotze committed:
Fix order of substitutions

jprotze committed:
Fix control_tool test. Runs now at any debug/optimization level

jprotze committed:
Add more tests for codeptr

jprotze committed:
Merge pull request #22 from hansangbae/towards_tr4

Declare omp_control_tool in Fortran header (Issue #21)

jprotze committed:
Fix non-OMPT build (missing OMPT_SUPPORT guard)

hansangbae committed:
Merge llvm-mirror/openmp manually

hansangbae committed:
Merge remote-tracking branch 'upstream_openmp/master' into towards_tr4

Conflicts:
	runtime/src/kmp.h
	runtime/src/kmp_atomic.h
	runtime/src/kmp_barrier.cpp
	runtime/src/kmp_cancel.cpp
	runtime/src/kmp_csupport.cpp
	runtime/src/kmp_dispatch.cpp
	runtime/src/kmp_ftn_entry.h
	runtime/src/kmp_ftn_os.h
	runtime/src/kmp_global.cpp
	runtime/src/kmp_gsupport.cpp
	runtime/src/kmp_lock.cpp
	runtime/src/kmp_lock.h
	runtime/src/kmp_runtime.cpp
	runtime/src/kmp_sched.cpp
	runtime/src/kmp_settings.cpp
	runtime/src/kmp_taskdeps.cpp
	runtime/src/kmp_tasking.cpp
	runtime/src/kmp_wait_release.h
	runtime/src/ompt-event-specific.h
	runtime/src/ompt-general.cpp
	runtime/src/ompt-internal.h
	runtime/src/ompt-specific.cpp
	runtime/src/ompt-specific.h

jprotze committed:
Merge pull request #23 from hansangbae/towards_tr4

Merge llvm-mirror/openmp master
This merges in the big reformating diff.

jprotze committed:
Add clang-format options to get consistent behavior.

jprotze committed:
Cleanup return/frame addresses for tasks:

- only kmpc-functions should invoke the built-in functions
- if a kmpc-function is called in the context of the runtime, both addresses need special handling (don't overwrite the value set on initial entry)
- Initial task has no return address
- Marking task as identified to be cancelled happens implicitly, no return address

jprotze committed:
The return address for OpenMP runtime routines that invoke callbacks needs to be stored on runtime entry.

jprotze committed:
Fix return address for serial parallel team and critical

jprotze committed:
Locking functions cannot store return-pointer in ompt_thread_info, if runtime is not initialized. Pass codeptr as an argument

jprotze committed:
Cleanup unnecessary LOAD/STORE of codeptr, fix codeptr for master barrier

jprotze committed:
Activate OMPT by default for non-Win/non-Darwin systems

jprotze committed:
Remove superflous include. Also fixes non-OMPT builts

jprotze committed:
Check return address for other OpenMP locking functions

jprotze committed:
Replace parallel_function in test cases by codeptr_ra, since that's what it really is

jprotze committed:
Pass return-address from runtime entry to the callback for worksharing loops.

jprotze committed:
Remove ancient and unused tracking of parallel/task-function.

jprotze committed:
Remove unused event from test callback.h

jprotze committed:
Provide typedefs and types for target-related callbacks, but no implementation.
This is also signaled to the tool when registering one of these callbacks.

jprotze committed:
Fetch codeptr for loop event on runtime entry

omalyshe committed:
Fixed build on Mac OS X

jprotze committed:
Verified that OMPT specific code for tasks is only active if there is a tool.

jprotze committed:
Merge pull request #28 from omalyshe/towards_tr4

Fixed build on Mac OS X

hansangbae committed:
Some cleanup in the tool activation, support for Windows.

hansangbae committed:
Merge remote-tracking branch 'upstream_master/master' into towards_tr4

Conflicts:
	runtime/src/kmp_tasking.cpp

hansangbae committed:
Add __declspec(dllexport) to ompt_start_tool's declaration for Windows

jprotze committed:
Make ompt_enabled a bitfield

This saves to load the ompt_callbacks structure with callback pointers for
the comparison. This should save some cache-lines in case the callback is
not registered.

hansangbae committed:
Fix incorrect changes in c0dc520

jprotze committed:
Remove commented code

jprotze committed:
Use __builtin_expect to trick branch prediction for OMPT

Assume that there is no tool by default. If there is a tool, the function call is more expensive than the missed branch prediction.

jprotze committed:
merge ompt-specific blocks for included tasks

hansangbae committed:
Merge remote-tracking branch 'upstream/towards_tr4' into towards_tr4

hansangbae committed:
Fix performance regression in 376.kdtree

jprotze committed:
Merge pull request #30 from hansangbae/towards_tr4

Tool initialization on Windows
Reduce runtime overhead for undeferred tasks

jprotze committed:
Prefer the use of weak attribute for ompt_start_tool

sconvent committed:
fix parallel_end return address for GOMP interface

sconvent committed:
fix return addresses for ordered and critical constructs using GOMP interface

sconvent committed:
fix untied flag for GOMP interface

sconvent committed:
load intel license file when available

sconvent committed:
fixes for return addresses for KMP interface for the ordered and critical constructs

sconvent committed:
fix compatibility issue with icc and clang

sconvent committed:
add stub for ompt_get_target_info

sconvent committed:
test whether a thread is an OpenMP thread in ompt_get_thread_data

sconvent committed:
add split worksharing testcases to test ompt return addresses

jprotze committed:
Return value of initialize function determines whether OMPT is enabled

jprotze committed:
Calling OpenMP runtime routine from callback is unsafe. Assume thread 0

sconvent committed:
replace critical section in ompt-signal with atomic construct

sconvent committed:
fix comment in ompt_get_target_info stub

sconvent committed:
better match invoker in ompt testcases

sconvent committed:
Check if parallel_data is not null for implicit parallel region

sconvent committed:
Compatibility fixes for icc

jprotze committed:
Fix redefinition.

sconvent committed:
Add missing OMPT guards

jprotze committed:
Tasks executed in serial context should not claim parallel execution

jprotze committed:
Fix runtime entry points: start handling NULL-pointer

jprotze committed:
Fix omp atomic for tests

sconvent committed:
Introduce fuzzy testing for return addresses for _split testcases

sconvent committed:
Further checks to avoid segfaults in runtime entry points

sconvent committed:
Handle different knid of loop scheduling in _split tests

sconvent committed:
Handle different kind of loop scheduling in _split tests

sconvent committed:
Merge branch 'towards_tr4' of git.rwth-aachen.de:OpenMPTools/LLVM-openmp into towards_tr4

erges an updated upstream into a topic branch.

sconvent committed:
Fix some return addresses for GOMP interface

sconvent committed:
Store KMP and GOMP return address in same variable

omalyshe committed:
Implemented runtime support for OMPT worksharing (Issue #6)

sconvent committed:
Workaround for icc bug in testcases

sconvent committed:
Fix testcases to be able to use older C standards

sconvent committed:
Null-check internal thread data to avoid segmentation faults during asynchronous calls

sconvent committed:
Add note to runtime entry point that may not be safe for asynchronous calls

sconvent committed:
Add note explaining a compiler bug workaround

sconvent committed:
Make sure the runtime is initialized on call of print_frame()

sconvent committed:
Fix return address storing in ompt_control_tool callback

sconvent committed:
Test return address in control_tool testcase

sconvent committed:
Add stub for ompt_get_num_devices

sconvent committed:
Cleanup tool_available testcase

sconvent committed:
Remove libunwind and backtrace code

sconvent committed:
Use fuzzy addresses in more testcases to achieve Intel compiler compatibility

sconvent committed:
Divide worksharing split testcases into two parts to be able to match events regardless of scheduling strategy

jprotze committed:
Revert 31483d0, fix the implementation of codeptr_ra for the init runtime routines

jprotze committed:
Missed some changes on merge

jprotze committed:
Fix portability issue.

omalyshe committed:
Merge remote-tracking branch 'upstream/towards_tr4' into towards_tr4

omalyshe committed:
Fix return address store in __kmpc_taskloop

jprotze committed:
Merge pull request #35 from omalyshe/towards_tr4

Implemented runtime support for OMPT worksharing (Issue #6)

jprotze committed:
Merge branch 'master' into towards_tr4

Commit:
Remove unused t_single_lock

Add padding inside team structure to keep same structure size.

git-svn-id: https://llvm.org/svn/llvm-project/openmp/trunk@314242 91177308-0d34-0410-b5e6-96231b3b80d8

Commit:
Allow printing of KMP_TOPOLOGY_METHOD when KMP_SETTINGS=true

git-svn-id: https://llvm.org/svn/llvm-project/openmp/trunk@314243 91177308-0d34-0410-b5e6-96231b3b80d8

Commit:
Remove unnecessary semicolons

Removes semicolons after if {} blocks, function definitions, etc.
I was able to apply the large OMPT patch cleanly on top of this one
with no conflicts.

git-svn-id: https://llvm.org/svn/llvm-project/openmp/trunk@314340 91177308-0d34-0410-b5e6-96231b3b80d8

sconvent committed:
Remove unnecessary sort-thread statements in ompt testcases

sconvent committed:
Test return address in all ompt parallel testcases

sconvent committed:
Fix ompt master testcase

sconvent committed:
Test return addresses in ompt task testcases

sconvent committed:
Test return addresses in ompt cancel_parallel testcase

sconvent committed:
Fix ompt nest_lock testcase

sconvent committed:
Test return addresses in ompt barrier testcases

sconvent committed:
Fix task_create callbacks for GOMP interface

sconvent committed:
Use shorter name for load_tool testcase

sconvent committed:
Use shorter name for load_tool testcase

hahnjo committed:
[test] Fix uninitialized memory in omp_taskloop_grainsize.c

result was never initialized to zero which sometimes failed the test.

git-svn-id: https://llvm.org/svn/llvm-project/openmp/trunk@314513 91177308-0d34-0410-b5e6-96231b3b80d8

sconvent committed:
Merge branch 'towards_tr4' of git.rwth-aachen.de:OpenMPTools/LLVM-openmp into towards_tr4

Conflicts:
	runtime/test/ompt/loadtool/tool.c

sconvent committed:
Revert "Activate OMPT by default for non-Win/non-Darwin systems"

This reverts commit 5e8481768c10f2fa1825dffb3386589eb6cd7615.

sconvent committed:
delete README.md

sconvent committed:
Move OpenMP version check for OMPT to correct file

sconvent committed:
Remove old ompt_control entry from exports_so.txt

sconvent committed:
Cleanup Fortran omp definitions

sconvent committed:
Cleanup and add comments in ompt.h

sconvent committed:
Cleanup in kmp_barrier.cpp

sconvent committed:
Cleanup in kmp_csupport.cpp

sconvent committed:
Cleanup in kmp_dispatch.cpp

sconvent committed:
Cleanup in kmp_ftn_entry.h and kmp_global.cpp

sconvent committed:
Cleanup kmp_gsupport.cpp

sconvent committed:
Cleanup in kmp_runtime.cpp

sconvent committed:
Cleanup in kmp_sched.cpp and kmp_settings.cpp

sconvent committed:
Cleanup in kmp_tasking.cpp

sconvent committed:
Cleanup in kmp_sched.cpp,  kmp_wait_release.h and ompt-general.cpp

sconvent committed:
Cleanup in ompt-specific.cpp and ompt-specific.h

sconvent committed:
Revert 22f75144 and fix issues

Commit:
KMP_HW_SUBSET vs KMP_PLACE_THREADS rival envirables fix

If both KMP_HW_SUBSET and KMP_PLACE_THREADS are set and KMP_PLACE_THREADS gets
parsed first, then the current environment variable parser rejects both and
neither get used. This patch uses the rivals mechanism that is used for other
environment variable groups (e.g., KMP_STACKSIZE, GOMP_STACKSIZE, OMP_STACKSIZE).
If both are set, then it tells the user that it is ignoring KMP_PLACE_THREADS in
favor of KMP_HW_SUBSET. The message about deprecating KMP_PLACE_THREADS when it
is set is still printed regardless.

Differential Revision: https://reviews.llvm.org/D38292

git-svn-id: https://llvm.org/svn/llvm-project/openmp/trunk@315091 91177308-0d34-0410-b5e6-96231b3b80d8

jprotze committed:
Implement changes requested in phabricator review

jprotze committed:
Applying clang-format-3.8.1

jprotze committed:
Remove gomp versioning/aliasing for omp_control_tool

sconvent committed:
Move omp_control_tool definition in correct guards

sconvent committed:
Merge branch 'towards_tr4' of git.rwth-aachen.de:OpenMPTools/LLVM-openmp into towards_tr4

sconvent committed:
Add control_tool_no_ompt_support testcase

sconvent committed:
Remove commented code and cleanup

sconvent committed:
Cleanup in ompt testcases

sconvent committed:
Only call OMPT_STORE_RETURN_ADDRESS when ompt is enabled

sconvent committed:
Further cleanup

sconvent committed:
Remove commented code

sconvent committed:
Remove commented code

sconvent committed:
Merge branch 'towards_tr4' of git.rwth-aachen.de:OpenMPTools/LLVM-openmp into towards_tr4

jprotze committed:
Remove more remaining TODOs

jprotze committed:
Merge the OMPT load library test

jprotze committed:
strip superflous includes

jprotze committed:
Test template based versioning for OMPT taskwait

Commit:
Tweak template to inline in non-OMPT case

sconvent committed:
Apply templating and inlining to __kmpc_omp_task_begin_if0

sconvent committed:
Add OMPT_SUPPORT guard

sconvent committed:
Apply templating and inlining to __kmpc_omp_task_complete_if0

sconvent committed:
Remove commented code

sconvent committed:
Add OMPT_OPTIONAL guard

sconvent committed:
Fix ompt return addresses for GOMP interface

sconvent committed:
Use XFAIL to disable unsupported testcases for gcc

jprotze committed:
Merge branch 'master' into towards_tr4

Conflicts:
	runtime/.clang-format
	runtime/CMakeLists.txt
	runtime/src/kmp_barrier.cpp
	runtime/src/kmp_i18n.cpp
	runtime/src/kmp_runtime.cpp
	runtime/test/lit.cfg

jprotze committed:
Remove include as suggested by Jonas

jprotze committed:
Fix the return of void.

jprotze committed:
Do not inline for the taskwait function.

jprotze committed:
Implement suggestions by Jonas

jprotze committed:
Address comments by Jonas

jprotze committed:
Replace builtin_expect by UNLIKELY

jprotze committed:
Remove the forward declaration, remove some includes

jprotze commented:
This branch landed in master.
